# **Модуль 2. Домашнее задание 1**

1. Импортируйте библиотеку Pandas и покажите версию Pandas

In [ ]:
import pandas as pd

print(pd.__version__)

2.2.2


2. Импортируйте файл в датафрейм под названием "Electric_Vehicle_Population_Data.csv" из https://drive.google.com/file/d/172oRlGRIHck8CcuOHmLszw63VsfvigFg/view?usp=drive_link (вам необходимо скачать) и посмотрите информацию о датафрейме

In [ ]:
df = pd.read_csv('Electric_Vehicle_Population_Data.csv')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 210165 entries, 0 to 210164
Data columns (total 17 columns):
 #   Column                                             Non-Null Count   Dtype  
---  ------                                             --------------   -----  
 0   VIN (1-10)                                         210165 non-null  object 
 1   County                                             210161 non-null  object 
 2   City                                               210161 non-null  object 
 3   State                                              210165 non-null  object 
 4   Postal Code                                        210161 non-null  float64
 5   Model Year                                         210165 non-null  int64  
 6   Make                                               210165 non-null  object 
 7   Model                                              210165 non-null  object 
 8   Electric Vehicle Type                              210165 non-null  object

3. Сколько строк и столбцов имеет датафрейм?

Ответ: 17 столбцов и 210165 строк


4. Какие столбцы имеют пустые ячейки?

Ответ: County, City, Postal Code, Electric Range, Base MSRP, Legislative District, Vehicle Location, Electric Utility, 2020 Census Tract


5. Покажите верхние 6 строк датафрейма?

In [ ]:
df.head(6)

VIN (1-10)     County       City State  Postal Code  Model Year     Make  \
0  5UXTA6C0XM     Kitsap    Seabeck    WA      98380.0        2021      BMW   
1  5YJ3E1EB1J     Kitsap    Poulsbo    WA      98370.0        2018    TESLA   
2  WP0AD2A73G  Snohomish    Bothell    WA      98012.0        2016  PORSCHE   
3  5YJ3E1EB5J     Kitsap  Bremerton    WA      98310.0        2018    TESLA   
4  1N4AZ1CP3K       King    Redmond    WA      98052.0        2019   NISSAN   
5  3FA6P0PU5F  Snohomish    Bothell    WA      98012.0        2015     FORD   

      Model                   Electric Vehicle Type  \
0        X5  Plug-in Hybrid Electric Vehicle (PHEV)   
1   MODEL 3          Battery Electric Vehicle (BEV)   
2  PANAMERA  Plug-in Hybrid Electric Vehicle (PHEV)   
3   MODEL 3          Battery Electric Vehicle (BEV)   
4      LEAF          Battery Electric Vehicle (BEV)   
5    FUSION  Plug-in Hybrid Electric Vehicle (PHEV)   

  Clean Alternative Fuel Vehicle (CAFV) Eligibility  Electric Range  \
0           Clean Alternative Fuel Vehicle Eligible            30.0   
1           Clean Alternative Fuel Vehicle Eligible           215.0   
2             Not eligible due to low battery range            15.0   
3           Clean Alternative Fuel Vehicle Eligible           215.0   
4           Clean Alternative Fuel Vehicle Eligible           150.0   
5             Not eligible due to low battery range            19.0   

   Base MSRP  Legislative District  DOL Vehicle ID  \
0        0.0                  35.0       267929112   
1        0.0                  23.0       475911439   
2        0.0                   1.0       101971278   
3        0.0                  23.0       474363746   
4        0.0                  45.0       476346482   
5        0.0                  21.0         9620047   

                  Vehicle Location  \
0  POINT (-122.8728334 47.5798304)   
1  POINT (-122.6368884 47.7469547)   
2    POINT (-122.206146 47.839957)   
3  POINT (-122.6231895 47.5930874)   
4      POINT (-122.13158 47.67858)   
5    POINT (-122.206146 47.839957)   

                                Electric Utility  2020 Census Tract  
0                         PUGET SOUND ENERGY INC       5.303509e+10  
1                         PUGET SOUND ENERGY INC       5.303509e+10  
2                         PUGET SOUND ENERGY INC       5.306105e+10  
3                         PUGET SOUND ENERGY INC       5.303508e+10  
4  PUGET SOUND ENERGY INC||CITY OF TACOMA - (WA)       5.303303e+10  
5                         PUGET SOUND ENERGY INC       5.306104e+10

6. Покажите нижние 3 строк датафрейма?

In [ ]:
df.tail(3)

VIN (1-10)   County              City State  Postal Code  Model Year  \
210162  JTMAB3FV3P   Pierce  University Place    WA      98466.0        2023   
210163  5YJ3E1EA1M     King          Kirkland    WA      98034.0        2021   
210164  7SAYGDEE0P  Spokane           Spokane    WA      99208.0        2023   

          Make       Model                   Electric Vehicle Type  \
210162  TOYOTA  RAV4 PRIME  Plug-in Hybrid Electric Vehicle (PHEV)   
210163   TESLA     MODEL 3          Battery Electric Vehicle (BEV)   
210164   TESLA     MODEL Y          Battery Electric Vehicle (BEV)   

        Clean Alternative Fuel Vehicle (CAFV) Eligibility  Electric Range  \
210162            Clean Alternative Fuel Vehicle Eligible            42.0   
210163  Eligibility unknown as battery range has not b...             0.0   
210164  Eligibility unknown as battery range has not b...             0.0   

        Base MSRP  Legislative District  DOL Vehicle ID  \
210162        0.0                  28.0       254488797   
210163        0.0                   1.0       137279852   
210164        0.0                   6.0       241306740   

                       Vehicle Location  \
210162   POINT (-122.538546 47.2298959)   
210163  POINT (-122.2026532 47.7210518)   
210164  POINT (-117.4268937 47.7323627)   

                                         Electric Utility  2020 Census Tract  
210162  BONNEVILLE POWER ADMINISTRATION||CITY OF TACOM...       5.305307e+10  
210163      PUGET SOUND ENERGY INC||CITY OF TACOMA - (WA)       5.303302e+10  
210164  BONNEVILLE POWER ADMINISTRATION||AVISTA CORP||...       5.306301e+10

7. Заполните пустые ячейки столбца Legislative District максимальным значением данного столбца

In [ ]:
df['Legislative District'].fillna(df['Legislative District'].max(), inplace=True)

<ipython-input-4-28095eead14e>:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['Legislative District'].fillna(df['Legislative District'].max(), inplace=True)


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 210165 entries, 0 to 210164
Data columns (total 17 columns):
 #   Column                                             Non-Null Count   Dtype  
---  ------                                             --------------   -----  
 0   VIN (1-10)                                         210165 non-null  object 
 1   County                                             210165 non-null  object 
 2   City                                               210161 non-null  object 
 3   State                                              210165 non-null  object 
 4   Postal Code                                        210161 non-null  float64
 5   Model Year                                         210165 non-null  int64  
 6   Make                                               210165 non-null  object 
 7   Model                                              210165 non-null  object 
 8   Electric Vehicle Type                              210165 non-null  object

8. Заполните пустые ячейки столбца County значением 'No data'

In [ ]:
df['County'].fillna('No data', inplace=True)

<ipython-input-5-79edf854f807>:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['County'].fillna('No data', inplace=True)


9. Заполните пустые ячейки столбца  Electric Range средним значением  столбца Base MSRP

In [ ]:
df['Electric Range'].fillna(df['Base MSRP'].mean(), inplace=True)

<ipython-input-6-6d77dc0e3f14>:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['Electric Range'].fillna(df['Base MSRP'].mean(), inplace=True)


10. Теперь после заполнения пустых ячеек по 2-м вышеуказанным столбцам, неоходимо удалить все строки, которые содержат пустые ячейка

In [ ]:
df.dropna(inplace=True)

11. Используя индексы датафрейма с 20 по 150, вытащите данные по столбцам 'VIN (1-10)', 'County', 'City', 'State', 'Postal Code', 'Model Year',
       'Make', 'Model', 'Electric Vehicle Type'. Результат запишите на новый датафрейм

In [ ]:
newdf = df.loc[[20, 150],
               ['VIN (1-10)', 'County', 'City', 'State', 'Postal Code', 'Model Year', 'Make', 'Model', 'Electric Vehicle Type']]


12. Покажите кол-во строк и столбцов используя .shape

In [ ]:
df.shape


(210150, 17)

13. Используя функцию Pandas покажите статистиечкие результаты основного датафрейма

In [ ]:
df.describe()


Postal Code     Model Year  Electric Range      Base MSRP  \
count  210150.000000  210150.000000   210150.000000  210150.000000   
mean    98178.194647    2021.048670       50.602579     897.312039   
std      2445.491444       2.988946       86.974319    7652.606144   
min      1731.000000    1999.000000        0.000000       0.000000   
25%     98052.000000    2019.000000        0.000000       0.000000   
50%     98125.000000    2022.000000        0.000000       0.000000   
75%     98373.000000    2023.000000       42.000000       0.000000   
max     99577.000000    2025.000000      337.000000  845000.000000   

       Legislative District  DOL Vehicle ID  2020 Census Tract  
count         210150.000000    2.101500e+05       2.101500e+05  
mean              28.972510    2.290765e+08       5.297929e+10  
std               14.921063    7.115445e+07       1.551507e+09  
min                1.000000    4.469000e+03       1.001020e+09  
25%               17.000000    1.948825e+08       5.303301e+10  
50%               33.000000    2.405161e+08       5.303303e+10  
75%               42.000000    2.629754e+08       5.305307e+10  
max               49.000000    4.792548e+08       5.602100e+10

14. Из полученного стат результата, какой столбец имеет 75% значение 0?

Ответ: Base MSRP

15. Покажите корреляцию между столбцами 'Base MSRP', 'Electric Range'

In [ ]:
df[['Base MSRP', 'Electric Range']].corr()

Base MSRP  Electric Range
Base MSRP        1.000000        0.114177
Electric Range   0.114177        1.000000

16. Что думаете о результате корреляции?

Ответ: если коррелция 1 - это отличный показатель, при увеличение одного значения другое значение тоже будет увелич.. Однако если значение корреляции 0.11 - это может говорить о том, что зависимость между переменными незначтельная. Если увеличить одно значние, другое будет незначетельно увеличиваться.

17.  Необходимо отсортировать от меньшего к большеми по столбцу Model Year и перезапишите датафреим

In [ ]:
df = df.sort_values(by='Model Year', ascending=True)

18. Создайте функцию, где смотря столбец 'Electric Range', необходимо заполнить новый столбец 'Electric Range Mapped' по условиям, если 'Electric Range' меньше 100 то 'Electric Range Mapped' равно 'Low', от 100 до 250 'Medium', выше 250 'High'

In [ ]:
def Electric_Range_Map(value):
  if value < 100:
    return 'Low'
  elif 100 <= value <= 250:
    return 'Medium'
  else:
    return 'High'

df['Electric Range Mapped'] = df['Electric Range'].map(Electric_Range_Map)


In [ ]:
df[['Electric Range','Electric Range Mapped']]

Electric Range Electric Range Mapped
0                 30.0                   Low
1                215.0                Medium
2                 15.0                   Low
3                215.0                Medium
4                150.0                Medium
...                ...                   ...
210160            26.0                   Low
210161            26.0                   Low
210162            42.0                   Low
210163             0.0                   Low
210164             0.0                   Low

[210150 rows x 2 columns]

In [ ]:
df[df['Electric Range Mapped'] == 'High']

VIN (1-10)     County           City State  Postal Code  Model Year  \
6       5YJYGDEEXL       King         Renton    WA      98055.0        2020   
13      5YJSA1DPXC   Thurston        Olympia    WA      98502.0        2012   
15      5YJYGDEF8L  Snohomish        Edmonds    WA      98026.0        2020   
16      5YJ3E1EA6L       King         Duvall    WA      98019.0        2020   
26      5YJYGDEE2L  Snohomish       Lynnwood    WA      98036.0        2020   
...            ...        ...            ...   ...          ...         ...   
210102  5YJYGDEE9L    Whatcom     Bellingham    WA      98226.0        2020   
210108  5YJYGDEE8L     Chelan         Chelan    WA      98816.0        2020   
210128  5YJYGDEE9L   San Juan  Friday Harbor    WA      98250.0        2020   
210153  1G1FZ6S07L   San Juan   Lopez Island    WA      98261.0        2020   
210154  KM8K23AGXL       King        Seattle    WA      98103.0        2020   

             Make    Model           Electric Vehicle Type  \
6           TESLA  MODEL Y  Battery Electric Vehicle (BEV)   
13          TESLA  MODEL S  Battery Electric Vehicle (BEV)   
15          TESLA  MODEL Y  Battery Electric Vehicle (BEV)   
16          TESLA  MODEL 3  Battery Electric Vehicle (BEV)   
26          TESLA  MODEL Y  Battery Electric Vehicle (BEV)   
...           ...      ...                             ...   
210102      TESLA  MODEL Y  Battery Electric Vehicle (BEV)   
210108      TESLA  MODEL Y  Battery Electric Vehicle (BEV)   
210128      TESLA  MODEL Y  Battery Electric Vehicle (BEV)   
210153  CHEVROLET  BOLT EV  Battery Electric Vehicle (BEV)   
210154    HYUNDAI     KONA  Battery Electric Vehicle (BEV)   

       Clean Alternative Fuel Vehicle (CAFV) Eligibility  Electric Range  \
6                Clean Alternative Fuel Vehicle Eligible           291.0   
13               Clean Alternative Fuel Vehicle Eligible           265.0   
15               Clean Alternative Fuel Vehicle Eligible           291.0   
16               Clean Alternative Fuel Vehicle Eligible           266.0   
26               Clean Alternative Fuel Vehicle Eligible           291.0   
...                                                  ...             ...   
210102           Clean Alternative Fuel Vehicle Eligible           291.0   
210108           Clean Alternative Fuel Vehicle Eligible           291.0   
210128           Clean Alternative Fuel Vehicle Eligible           291.0   
210153           Clean Alternative Fuel Vehicle Eligible           259.0   
210154           Clean Alternative Fuel Vehicle Eligible           258.0   

        Base MSRP  Legislative District  DOL Vehicle ID  \
6             0.0                  11.0       124565731   
13        59900.0                  22.0       188634442   
15            0.0                  21.0       112172084   
16            0.0                  45.0       107653255   
26            0.0                   1.0       183054350   
...           ...                   ...             ...   
210102        0.0                  42.0       124333422   
210108        0.0                  12.0       195946908   
210128        0.0                  40.0       114738186   
210153        0.0                  40.0       192490203   
210154        0.0                  43.0       241649292   

                       Vehicle Location  \
6       POINT (-122.2003346 47.4487206)   
13          POINT (-122.92333 47.03779)   
15        POINT (-122.338942 47.821454)   
16        POINT (-121.956372 47.735604)   
26          POINT (-122.29245 47.82557)   
...                                 ...   
210102       POINT (-122.49756 48.7999)   
210108  POINT (-120.0453013 47.8563197)   
210128  POINT (-123.0679454 48.5327669)   
210153  POINT (-122.8826643 48.5053168)   
210154   POINT (-122.3499053 47.673887)   

                                         Electric Utility  2020 Census Tract  \
6           PUGET SOUND ENERGY INC||CITY OF TACOMA - (WA)       5.303303e+10   
13             

19. Покажите все строки индекса 1532 с помощью метода .loc

In [ ]:
df.loc[1532]

VIN (1-10)                                                                                  KM8KNDA80S
County                                                                                            King
City                                                                                            Burien
State                                                                                               WA
Postal Code                                                                                    98168.0
Model Year                                                                                        2025
Make                                                                                           HYUNDAI
Model                                                                                        IONIQ 5 N
Electric Vehicle Type                                                   Battery Electric Vehicle (BEV)
Clean Alternative Fuel Vehicle (CAFV) Eligibility    Eligibility unknown as battery range has not b...
Electric Range                                                                                     0.0
Base MSRP                                                                                          0.0
Legislative District                                                                              33.0
DOL Vehicle ID                                                                               270066149
Vehicle Location                                                         POINT (-122.297205 47.486649)
Electric Utility                                          CITY OF SEATTLE - (WA)|CITY OF TACOMA - (WA)
2020 Census Tract                                                                        53033027000.0
Name: 1532, dtype: object

20. Сохраните датафреим на формат csv и xlsx(Excel). Наименование файла можете дать по желанию

In [ ]:
df.to_csv('electric_vech_populat.csv', index=False)

In [ ]:
df.to_csv('electric_vech_populat.xlsx', index=False)